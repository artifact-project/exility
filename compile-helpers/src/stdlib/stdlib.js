"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
var R_HTML_ENTITES = /[&<>"]/;
var R_HTML_ENTITY_AMP = /&/g;
var R_HTML_ENTITY_LT = /</g;
var R_HTML_ENTITY_GT = />/g;
var R_HTML_ENTITY_QUOTE = /"/g;
function NOOP() { }
exports.default = {
    NIL: null,
    R_HTML_ENTITES: R_HTML_ENTITES,
    R_HTML_ENTITY_AMP: R_HTML_ENTITY_AMP,
    R_HTML_ENTITY_LT: R_HTML_ENTITY_LT,
    R_HTML_ENTITY_GT: R_HTML_ENTITY_GT,
    R_HTML_ENTITY_QUOTE: R_HTML_ENTITY_QUOTE,
    NOOP: NOOP,
    TO_STRING: function (value) {
        return value == null ? '' : value + '';
    },
    RETURN_EMPTY_STRING: function () {
        return '';
    },
    HTML_ENCODE: function (value) {
        if (value == null) {
            return '';
        }
        if (R_HTML_ENTITES.test(value)) {
            if (value.indexOf('&') !== -1) {
                value = value.replace(R_HTML_ENTITY_AMP, '&amp;');
            }
            if (value.indexOf('<') !== -1) {
                value = value.replace(R_HTML_ENTITY_LT, '&lt;');
            }
            if (value.indexOf('>') !== -1) {
                value = value.replace(R_HTML_ENTITY_GT, '&gt;');
            }
            if (value.indexOf('"') !== -1) {
                value = value.replace(R_HTML_ENTITY_QUOTE, '&quot;');
            }
            return value;
        }
        return value.toString();
    },
    EACH: function (data, callback) {
        if (data != null) {
            if (data.forEach) {
                var length_1 = data.length;
                for (var i = 0; i < length_1; i++) {
                    callback(data[i], i);
                }
            }
            else {
                for (var key in data) {
                    if (data.hasOwnProperty(key)) {
                        callback(data[key], key);
                    }
                }
            }
        }
    },
    SLOT: function (parent, slots, __super__, name, fn) {
        if (slots && slots.hasOwnProperty(name)) {
            slots[name](parent, __super__);
        }
        else {
            fn();
        }
    },
    CLONE_OBJECT: function (target) {
        var cloned = {};
        if (target != null) {
            for (var key in target) {
                cloned[key] = target[key];
            }
        }
        return cloned;
    }
};
