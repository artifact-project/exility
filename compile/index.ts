import COMMON_TEST from './src/COMMON_TEST/COMMON_TEST';
import SELF_CLOSED_TAGS from './src/SELF_CLOSED_TAGS/SELF_CLOSED_TAGS';

import createCompiler, {ICompilerOptions} from './src/createCompiler/createCompiler';

import stringifyObjectKey from './src/stringifyObjectKey/stringifyObjectKey';
import stringifyParsedValue from './src/stringifyParsedValue/stringifyParsedValue';
import stringifyAttributeValue from './src/stringifyAttributeValue/stringifyAttributeValue';
import simpleJavaScriptBeautifier from './src/simpleJavaScriptBeautifier/simpleJavaScriptBeautifier';


export {
	COMMON_TEST,
	SELF_CLOSED_TAGS,

	createCompiler,
	ICompilerOptions,

	stringifyObjectKey,
	stringifyParsedValue,
	stringifyAttributeValue,
	simpleJavaScriptBeautifier,
};

export function applyTraits(traits, raw) {
	raw.traits && raw.traits.forEach(({name}) => {
		if (traits.hasOwnProperty(name)) {
			const trait = traits[name];

			for (let key in trait) {
				if (trait.hasOwnProperty(key)) {
					if (raw.attrs.hasOwnProperty(key)) {
						raw.attrs[key] = trait[key].concat(raw.attrs[key]);
					} else {
						raw.attrs[key]= trait[key];
					}
				}
			}
		}
	});
}
