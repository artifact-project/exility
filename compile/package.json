{
  "name": "@exility/compile",
  "version": "0.1.0",
  "description": "Exility compile helpers",
  "main": "index.js",
  "scripts": {
    "test": "jest --coverage",
    "ts-build": "tsc",
    "git-clean": "git clean -f",
    "prepublishOnly": "npm run ts-build",
    "postpublish": "npm run git-clean"
  },
  "files": [
    "*.js",
    "*.d.ts",
    "src/*/*.js",
    "src/*/*.d.ts"
  ],
  "author": "RubaXa <ibnRubaXa@gmail.com>",
  "license": "MIT",
  "dependencies": {
    "tslib": "^1.9.3"
  },
  "devDependencies": {
    "@exility/block": "0.1.0",
    "@exility/parser": "0.1.0",
    "@types/jest": "^23.3.3",
    "jest": "^23.6.0",
    "ts-jest": "^23.10.4",
    "typescript": "^3.1.1"
  },
  "jest": {
    "transform": {
      "^.+\\.ts$": "<rootDir>/node_modules/ts-jest/preprocessor.js"
    },
    "testRegex": "\\.tests\\.ts$",
    "setupTestFrameworkScriptFile": "<rootDir>/__jest__/extensions.ts",
    "moduleFileExtensions": [
      "ts",
      "tsx",
      "js",
      "json"
    ]
  }
}
