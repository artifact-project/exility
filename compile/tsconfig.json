{
  "compilerOptions": {
    "module": "commonjs",
    "outDir": "dist/",
    "target": "es5",
    "declaration": true,
    "moduleResolution": "node",
    "experimentalDecorators": true,
	"lib": [
		"es5",
		"es2016.array.include"
	]
  },
  "exclude": ["dist", "node_modules"]
}
