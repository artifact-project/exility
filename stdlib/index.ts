import core from './src/core/core';
import dom from './src/dom/dom';

export {
	core,
	dom,
};
