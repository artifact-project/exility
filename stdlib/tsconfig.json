{
  "compilerOptions": {
    "module": "commonjs",
    "target": "es5",
    "declaration": true,
    "moduleResolution": "node",
    "experimentalDecorators": true,
    "importHelpers": true,
    "noEmitHelpers": true,
    "lib": [
      "dom",
      "es5",
      "es2015.promise",
      "es2016.array.include"
    ]
  },
  "exclude": [
    "node_modules"
  ]
}
