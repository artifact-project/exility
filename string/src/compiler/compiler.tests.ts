import {COMMON_TEST} from '@exility/compile';
import {core as stdlib} from '@exility/stdlib';
import compilerFactory from './compiler';

const fromString = function (input: string, scope?: string[], pure?: boolean) {
	const compiler = compilerFactory({
		debug: true,
		pure,
		scope,
	});

	return compiler(input);
};

it('doctype', () => {
	const factory = fromString(`!html`);

	expect(factory).toMatchSnapshot();
	expect(factory()()).toBe('<!DOCTYPE html>');
});

it('page / prettify', () => {
	const prettifyCompiler = compilerFactory({
		debug: true,
		prettify: true,
		scope: [],
	});
	const factory = prettifyCompiler(`html\n\thead > title | foo\n\tbody > h1.title | Bar`);

	expect(factory).toMatchSnapshot();
	expect(factory()()).toMatchSnapshot();
});

it('interpolate', () => {
	const factory = fromString('h1.title-${size} | Hi, ${user}!', ['user', 'size']);

	expect(factory).toMatchSnapshot();
	expect(factory({stdlib})({user: '%name%', size: 'xxl'})).toMatchSnapshot();
});

describe('meta-comments for isomorphic', () => {
	const withMetaCommentsCompiler = compilerFactory({
		scope: ['x', 'y'],
		metaComments: true,
	});

	it('value', () => {
		const template = withMetaCommentsCompiler('| ${x} ${y}?')({stdlib});

		expect(template({x: '', y: 'bar'})).toBe('<!--(--><!--)--> <!--(-->bar<!--)-->?');
		expect(template({x: 'foo', y: 'bar'})).toBe('<!--(-->foo<!--)--> <!--(-->bar<!--)-->?');
	});

	it('if', () => {
		const template = withMetaCommentsCompiler('if (x) > | \${y}?')({stdlib});

		expect(template({x: false})).toBe('<!--if1--><!--/if1-->');
		expect(template({x: true, y: 'ok'})).toBe('<!--if1--><!--(-->ok<!--)-->?<!--/if1-->');
	});

	it('for', () => {
		const template = withMetaCommentsCompiler('for (i in x) > | ${i}')({stdlib});

		expect(template({})).toBe('<!--for1--><!--/for1-->');
		expect(template({x: [1, 2]})).toBe('<!--for1--><!--(-->1<!--)--><!--(-->2<!--)--><!--/for1-->');
	});
});

it('html encoding', () => {
	const factory = fromString('| ${value}', ['value']);
	const template = factory({stdlib});
	const date = new Date();

	expect(factory).toMatchSnapshot();
	expect(template({})).toBe('');
	expect(template({value: null})).toBe('');
	expect(template({value: date})).toBe(date.toString());
	expect(template({value: 123})).toBe('123');
	expect(template({value: '<'})).toBe('&lt;');
	expect(template({value: '>'})).toBe('&gt;');
	expect(template({value: '&'})).toBe('&amp;');
	expect(template({value: '"'})).toBe('&quot;');
	expect(template({value: '<script>alert("&")</script>'})).toBe('&lt;script&gt;alert(&quot;&amp;&quot;)&lt;/script&gt;');
});

it('inherit', () => {
	const factory = fromString('.btn > .&__icon');
	expect(factory).toMatchSnapshot();
});

it('inherit + interpolate', () => {
	const factory = fromString('.${x} > .&__text', ['x']);

	expect(factory).toMatchSnapshot();
	expect(factory({stdlib})({x: 'btn'})).toMatchSnapshot();
});

it('panel = [title] + default slot', () => {
	const factory = fromString([
		'panel = [title]',
		'  h1 | ${title}',
		'  p > __default()',
		'panel[title="?!"]',
		'panel[title="Wow!"]',
		'  | Done',
	].join('\n'));

	expect(factory).toMatchSnapshot();
	expect(factory({stdlib})()).toMatchSnapshot();
});

it('panel = [title] + content slot', () => {
	const factory = fromString([
		'panel = [title]',
		'  content(title.toUpperCase(), "?")',
		'panel[title="wow!"]',
		'  content = (text, chr)',
		'    p | ${text}${chr}',
	].join('\n'));

	expect(factory({stdlib})()).toBe('<p>WOW!?</p>');
});

it('panel = [title] + content slot (+ default)', () => {
	const factory = fromString([
		'panel = [title]',
		'  content(title.toUpperCase(), "?")',
		'  content = (text, chr)',
		'    | ${text}${chr}',
		'h1 > panel[title="wow"]',
		'h2 > panel[title="xyz"]',
		'  content = (text)',
		'    | ${text.split("").reverse().join("")}',
	].join('\n'));

	expect(factory({stdlib})()).toBe('<h1>WOW?</h1><h2>ZYX</h2>');
});

it('panel = [title] + super', () => {
	const factory = fromString([
		'panel = [title]',
		'  content(title)',
		'  content = (text)',
		'    | ${text}',
		'panel[title="xyz"]',
		'  content = (text)',
		'    p > super.content(text.toUpperCase()) + | !'
	].join('\n'));

	expect(factory({stdlib})()).toBe('<p>XYZ!</p>');
});

it('pure', () => {
	const factory = fromString('i.foo | ${text}!', ['text'], true);

	expect(factory).toMatchSnapshot();
	expect(factory()({text: '&bar'})).toBe('<i class="foo">&amp;bar!</i>');
});

describe('COMMON_TEST', () => {
	COMMON_TEST.forEach(({title, template, data, results, snapshot}) => {
		results = [].concat(results);

		describe(title, () => {
			[].concat(data).forEach((scope, idx) => {
				const templateFactory = fromString(template, Object.keys(scope));

				it(JSON.stringify(scope), () => {
					if (!idx && snapshot) {
						expect(templateFactory).toMatchSnapshot();
					}

					expect(templateFactory({stdlib})(scope).replace(' href=""', '')).toBe(results[idx]);
				});
			});
		});
	});
});
