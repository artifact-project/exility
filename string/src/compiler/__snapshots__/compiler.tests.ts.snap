// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`COMMON_TEST Custom element {"text":"Wow!","href":null} 1`] = `
function templateFactory(__DEPS__) {
	"use strict";
	var __STDLIB = __DEPS__.stdlib;
	var __STDLIB_HTML_ENCODE = __STDLIB.HTML_ENCODE;
	function btn(attrs, __slots) {
		var __super = {attrs: attrs},
		text = attrs.text,
		url = attrs.url
		return "<" + __STDLIB_HTML_ENCODE(url ? "a" : "button") + " class=\\"btn\\" href=\\"" + __STDLIB_HTML_ENCODE(url) + "\\">" + __STDLIB_HTML_ENCODE(text) + "</" + __STDLIB_HTML_ENCODE(url ? "a" : "button") + ">";
	}

	return function compiledTemplate(__SCOPE__, __OPTIONS__) {
		__SCOPE__ = __SCOPE__ || {};
		var text = __SCOPE__.text;
		var href = __SCOPE__.href;

		return btn({text: (text), url: (href)});
	}
}
`;

exports[`COMMON_TEST FOR statement {"data":[1,2]} 1`] = `
function templateFactory(__DEPS__) {
	"use strict";
	var __STDLIB = __DEPS__.stdlib;
	var __STDLIB_EACH = __STDLIB.EACH;
	var __STDLIB_HTML_ENCODE = __STDLIB.HTML_ENCODE;
	return function compiledTemplate(__SCOPE__, __OPTIONS__) {
		__SCOPE__ = __SCOPE__ || {};
		var data = __SCOPE__.data;

		var __ROOT = "";
		__STDLIB_EACH(data, function EACH_ITERATOR(val, $index) {
			__ROOT += __STDLIB_HTML_ENCODE(val) + ",";
		});

		return __ROOT
	}
}
`;

exports[`COMMON_TEST FOR statement with key {"data":[1,2]} 1`] = `
function templateFactory(__DEPS__) {
	"use strict";
	var __STDLIB = __DEPS__.stdlib;
	var __STDLIB_EACH = __STDLIB.EACH;
	var __STDLIB_HTML_ENCODE = __STDLIB.HTML_ENCODE;
	return function compiledTemplate(__SCOPE__, __OPTIONS__) {
		__SCOPE__ = __SCOPE__ || {};
		var data = __SCOPE__.data;

		var __ROOT = "";
		__STDLIB_EACH(data, function EACH_ITERATOR(val, key) {
			__ROOT += __STDLIB_HTML_ENCODE(key) + ":" + __STDLIB_HTML_ENCODE(val) + ",";
		});

		return __ROOT
	}
}
`;

exports[`COMMON_TEST IF statement {"x":null} 1`] = `
function templateFactory() {
	"use strict";
	return function compiledTemplate(__SCOPE__, __OPTIONS__) {
		__SCOPE__ = __SCOPE__ || {};
		var x = __SCOPE__.x;

		var __ROOT = "<foo></foo>";
		if (x) {
			__ROOT += "<bar></bar>";
		}

		return __ROOT
	}
}
`;

exports[`COMMON_TEST IF/ELSE IF/ELSE statement {"x":1} 1`] = `
function templateFactory() {
	"use strict";
	return function compiledTemplate(__SCOPE__, __OPTIONS__) {
		__SCOPE__ = __SCOPE__ || {};
		var x = __SCOPE__.x;

		var __ROOT = "";
		if (x == 1) {
			__ROOT += "<a></a>";
		}else if (x == 2) {
			__ROOT += "<b></b>";
		}else {
			__ROOT += "<c></c>";
		}

		return __ROOT
	}
}
`;

exports[`COMMON_TEST IF/ELSE statement {"x":true} 1`] = `
function templateFactory() {
	"use strict";
	return function compiledTemplate(__SCOPE__, __OPTIONS__) {
		__SCOPE__ = __SCOPE__ || {};
		var x = __SCOPE__.x;

		var __ROOT = "";
		if (x) {
			__ROOT += "<a></a>";
		}else {
			__ROOT += "<b></b>";
		}

		return __ROOT
	}
}
`;

exports[`doctype 1`] = `
function templateFactory() {
	"use strict";
	return function compiledTemplate(__SCOPE__, __OPTIONS__) {
		return "<!DOCTYPE html>";
	}
}
`;

exports[`html encoding 1`] = `
function templateFactory(__DEPS__) {
	"use strict";
	var __STDLIB = __DEPS__.stdlib;
	var __STDLIB_HTML_ENCODE = __STDLIB.HTML_ENCODE;
	return function compiledTemplate(__SCOPE__, __OPTIONS__) {
		__SCOPE__ = __SCOPE__ || {};
		var value = __SCOPE__.value;

		return __STDLIB_HTML_ENCODE(value);
	}
}
`;

exports[`inherit + interpolate 1`] = `
function templateFactory(__DEPS__) {
	"use strict";
	var __STDLIB = __DEPS__.stdlib;
	var __STDLIB_HTML_ENCODE = __STDLIB.HTML_ENCODE;
	return function compiledTemplate(__SCOPE__, __OPTIONS__) {
		__SCOPE__ = __SCOPE__ || {};
		var x = __SCOPE__.x;

		return "<div class=\\"" + __STDLIB_HTML_ENCODE(x) + "\\"><div class=\\"" + __STDLIB_HTML_ENCODE(x) + "__text\\"></div></div>";
	}
}
`;

exports[`inherit + interpolate 2`] = `<div class="btn"><div class="btn__text"></div></div>`;

exports[`inherit 1`] = `
function templateFactory() {
	"use strict";
	return function compiledTemplate(__SCOPE__, __OPTIONS__) {
		return "<div class=\\"btn\\"><div class=\\"btn__icon\\"></div></div>";
	}
}
`;

exports[`interpolate 1`] = `
function templateFactory(__DEPS__) {
	"use strict";
	var __STDLIB = __DEPS__.stdlib;
	var __STDLIB_HTML_ENCODE = __STDLIB.HTML_ENCODE;
	return function compiledTemplate(__SCOPE__, __OPTIONS__) {
		__SCOPE__ = __SCOPE__ || {};
		var user = __SCOPE__.user;
		var size = __SCOPE__.size;

		return "<h1 class=\\"title-" + __STDLIB_HTML_ENCODE(size) + "\\">Hi, " + __STDLIB_HTML_ENCODE(user) + "!</h1>";
	}
}
`;

exports[`interpolate 2`] = `<h1 class="title-xxl">Hi, %name%!</h1>`;

exports[`page / prettify 1`] = `
function templateFactory() {
	"use strict";
	return function compiledTemplate(__SCOPE__, __OPTIONS__) {
		return "<html>\\n  <head>\\n    <title>foo</title>\\n  </head>\\n  <body>\\n    <h1 class=\\"title\\">Bar</h1>\\n  </body>\\n</html>\\n";
	}
}
`;

exports[`page / prettify 2`] = `
<html>
  <head>
    <title>foo</title>
  </head>
  <body>
    <h1 class="title">Bar</h1>
  </body>
</html>

`;

exports[`panel = [title] + default slot 1`] = `
function templateFactory(__DEPS__) {
	"use strict";
	var __STDLIB = __DEPS__.stdlib;
	var __STDLIB_HTML_ENCODE = __STDLIB.HTML_ENCODE;
	function panel(attrs, __slots) {
		var __super = {attrs: attrs},
		title = attrs.title,
		__default = __slots && __slots.__default || __super.__default
		var __ROOT = "<h1>" + __STDLIB_HTML_ENCODE(title) + "</h1><p>";
		if (__default) {
			__ROOT += __default.call(__super);
		}
		__ROOT += "</p>";
		return __ROOT
	}

	return function compiledTemplate(__SCOPE__, __OPTIONS__) {
		var __ROOT = panel({title: "?!"}) + panel({title: "Wow!"}, {__default: function __default() {
			return "Done";
			}});

			return __ROOT
		}
	}
`;

exports[`panel = [title] + default slot 2`] = `<h1>?!</h1><p></p><h1>Wow!</h1><p>Done</p>`;

exports[`pure 1`] = `
function templateFactory() {
	"use strict";
	var __STDLIB = new (function () {
		var HTML_ENCODE = function HTML_ENCODE(value) {
			if (value == null) {
				return '';
			}
			if (R_HTML_ENTITES.test(value)) {
				if (value.indexOf('&') !== -1) {
					value = value.replace(R_HTML_ENTITY_AMP, '&amp;');
				}
				if (value.indexOf('<') !== -1) {
					value = value.replace(R_HTML_ENTITY_LT, '&lt;');
				}
				if (value.indexOf('>') !== -1) {
					value = value.replace(R_HTML_ENTITY_GT, '&gt;');
				}
				if (value.indexOf('"') !== -1) {
					value = value.replace(R_HTML_ENTITY_QUOTE, '&quot;');
				}
				return value;
			}
			return value.toString();
		};
		var R_HTML_ENTITES = /[&<>"]/;
		var R_HTML_ENTITY_AMP = /&/g;
		var R_HTML_ENTITY_LT = /</g;
		var R_HTML_ENTITY_GT = />/g;
		var R_HTML_ENTITY_QUOTE = /"/g;
		this.HTML_ENCODE = HTML_ENCODE;
	});

	var __STDLIB_HTML_ENCODE = __STDLIB.HTML_ENCODE;
	return function compiledTemplate(__SCOPE__, __OPTIONS__) {
		__SCOPE__ = __SCOPE__ || {};
		var text = __SCOPE__.text;

		return "<i class=\\"foo\\">" + __STDLIB_HTML_ENCODE(text) + "!</i>";
	}
}
`;
